From 95c7aa7d911178bf94433e422ca26530d674a055 Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@nbd.name>
Date: Mon, 18 Nov 2024 16:56:13 +0530
Subject: [PATCH] QSDK:CP: ucode: add support for nl80211 vif radio mask

Used for multi-radio support

patch-dependency: None
patch-work: None

Link: https://github.com/openwrt/openwrt/commit/5db27664bc1ccbdbf2f80c8517d1b16bcfe6b626

Signed-off-by: Felix Fietkau <nbd@nbd.name>
Signed-off-by: Rajat Soni <quic_rajson@quicinc.com>
---
 include/linux/nl80211.h | 4 ++++
 lib/nl80211.c           | 3 ++-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/include/linux/nl80211.h b/include/linux/nl80211.h
index 5ee1a4e..6984e3b 100644
--- a/include/linux/nl80211.h
+++ b/include/linux/nl80211.h
@@ -2918,6 +2918,8 @@ enum nl80211_commands {
  *	nested item, it contains attributes defined in
  *	&enum nl80211_if_combination_attrs.
  *
+ * @NL80211_ATTR_VIF_RADIO_MASK: Bitmask of allowed radios (u32).
+ *	A value of 0 means all radios.
  * @NL80211_ATTR_HE_MUEDCA_PARAMS: MU-EDCA AC parameters for the
  *	%NL80211_CMD_UPDATE_HE_MUEDCA_PARAMS command.
  * @NL80211_ATTR_BEACON_TX_MODE: used to configure the beacon tx mode as
@@ -3537,6 +3539,8 @@ enum nl80211_attrs {
 	NL80211_ATTR_WIPHY_RADIOS,
 	NL80211_ATTR_WIPHY_INTERFACE_COMBINATIONS,
 
+	NL80211_ATTR_VIF_RADIO_MASK,
+
 	NL80211_ATTR_WIPHY_ANTENNA_GAIN,
 
 	NL80211_ATTR_HE_MUEDCA_PARAMS,
diff --git a/lib/nl80211.c b/lib/nl80211.c
index 501fa68..2cdeba2 100644
--- a/lib/nl80211.c
+++ b/lib/nl80211.c
@@ -848,7 +848,7 @@ static const uc_nl_nested_spec_t nl80211_attr_multi_hw_macs = {
 
 static const uc_nl_nested_spec_t nl80211_msg = {
 	.headsize = 0,
-	.nattrs = 129,
+	.nattrs = 130,
 	.attrs = {
 		{ NL80211_ATTR_4ADDR, "4addr", DT_U8, 0, NULL },
 		{ NL80211_ATTR_AIRTIME_WEIGHT, "airtime_weight", DT_U16, 0, NULL },
@@ -978,6 +978,7 @@ static const uc_nl_nested_spec_t nl80211_msg = {
 		{ NL80211_ATTR_MAX_AP_ASSOC_STA, "max_ap_assoc", DT_U16, 0, NULL },
 		{ NL80211_ATTR_SURVEY_INFO, "survey_info", DT_NESTED, 0, &nl80211_survey_info_nla },
 		{ NL80211_ATTR_WIPHY_RADIOS, "radios", DT_NESTED, DF_MULTIPLE|DF_AUTOIDX, &nl80211_wiphy_radio_nla },
+		{ NL80211_ATTR_VIF_RADIO_MASK, "vif_radio_mask", DT_U32, 0, NULL },
 		{ NL80211_ATTR_MULTI_HW_MACS, "multi_hw_macs", DT_NESTED, DF_MULTIPLE|DF_AUTOIDX, &nl80211_attr_multi_hw_macs },
 	}
 };
-- 
2.34.1

